### PR TITLE
Send Build Status Report Using GitHub Status API

### PR DESCRIPTION
**Pull request by saadmk11**
Sending build status `pending` and `suncess` or `failed` using GitHub Status API.
Will extend this for other services in the future (e.g: gitlab, bitbucket)

**TODO:**

- [ ] Add Tests.

### CODE REVIEW

File: readthedocs/builds/constants.py [hunk 1]
```diff
@@ -61,3 +61,37 @@
     LATEST,
     STABLE,
 )
+
+# GitHub Build Statuses
+GITHUB_BUILD_STATE_FAILURE = 'failure'
+GITHUB_BUILD_STATE_PENDING = 'pending'
+GITHUB_BUILD_STATE_SUCCESS = 'success'
+
+# GitHub Build Statuses
```
Comment (ericholscher) [inline readthedocs/builds/constants.py]:
```suggestion
# GitLab Build Statuses
```

File: readthedocs/builds/models.py [hunk 2]
```diff
@@ -732,6 +732,14 @@ def __str__(self):
     def get_absolute_url(self):
         return reverse('builds_detail', args=[self.project.slug, self.pk])
 
+    def get_full_url(self):
+        """Get full url including domain"""
+        full_url = '{domain}{absolute_url}'.format(
+            domain=settings.PUBLIC_API_URL,
```
Comment (ericholscher) [inline readthedocs/builds/models.py]: This isn't the right setting. 
```suggestion
            domain=settings.PRODUCTION_DOMAIN,
```

File: readthedocs/oauth/services/github.py [hunk 3]
```diff
@@ -311,6 +312,83 @@ def update_webhook(self, project, integration):
...
+                )
+                return False
+
+        # Catch exceptions with request or deserializing JSON
+        except (RequestException, ValueError):
+            log.exception(
+                'GitHub commit status creation failed for project: %s',
+                project,
+            )
+        else:
```
Comment (ericholscher) [inline readthedocs/oauth/services/github.py]: Having an `else` for an error case feels weird. Is there a reason we're doing it this way? Why not just have it as part of the `try` block after the `return`'s?

File: readthedocs/core/utils/__init__.py [hunk 4]
```diff
@@ -125,6 +128,10 @@ def prepare_build(
     options['soft_time_limit'] = time_limit
     options['time_limit'] = int(time_limit * 1.2)
 
+    if build:
```
Comment (ericholscher) [inline readthedocs/core/utils/__init__.py]: Should this check for the `version` being external before sending?

File: readthedocs/projects/tasks.py [hunk 5]
```diff
@@ -573,6 +577,16 @@ def run_build(self, docker, record):
 
         if self.build_env.failed:
             self.send_notifications(self.version.pk, self.build['id'])
+            # send build failure status to git Status API
+            self.send_build_status(
+                self.build['id'], BUILD_STATUS_FAILURE
+            )
+
+        if self.build_env.successful:
```
Comment (ericholscher) [inline readthedocs/projects/tasks.py]: Should this check for the version being external before sending, same as above?

File: readthedocs/builds/constants.py (hunk 1 above)
Comment (ericholscher) [reply readthedocs/builds/constants.py]: Also, these aren't used currently are they?

File: readthedocs/core/utils/__init__.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/core/utils/__init__.py]: The status can be for any commit, I think it would be cool to report the status in normal versions

File: readthedocs/projects/tasks.py [hunk 6]
```diff
@@ -1768,3 +1789,32 @@ def retry_domain_verification(domain_pk):
...
+    """
+    build = Build.objects.get(pk=build_pk)
+    version = build.version
+    project = build.project
+
+    # Send status reports for only External (pull/merge request) Versions.
+    if version.type != EXTERNAL:
+        return
+
+    if 'github.com' in project.repo:
```
Comment (stsewd) [inline readthedocs/projects/tasks.py]: I think you can get this from `project.remote_repository.account.provider`

File: readthedocs/projects/tasks.py [hunk 7]
```diff
@@ -1768,3 +1789,32 @@ def retry_domain_verification(domain_pk):
...
+        account = project.remote_repository.account
+        service = GitHubService(project.users.first(), account)
```
Comment (stsewd) [inline readthedocs/projects/tasks.py]: I think the user should be the one from  `project.remote_repository.users`

File: readthedocs/projects/tasks.py [hunk 8]
```diff
@@ -1768,3 +1789,32 @@ def retry_domain_verification(domain_pk):
...
+
+        # send Status report using the API.
+        service.send_build_status(
+            project, version.identifier, state,
```
Comment (stsewd) [inline readthedocs/projects/tasks.py]: Looks like you can just send the build object here. Project, identifier and url can be got from that.

File: readthedocs/builds/constants.py (hunk 1 above)
Comment (saadmk11) [reply readthedocs/builds/constants.py]: No, They are not but we will use it in future. should I remove them for now?

File: readthedocs/builds/models.py (hunk 2 above)
Comment (saadmk11) [reply readthedocs/builds/models.py]: `PRODUCTION_DOMAIN` does not include `http` / `https` schemes.

File: readthedocs/oauth/services/github.py (hunk 3 above)
Comment (saadmk11) [reply readthedocs/oauth/services/github.py]: We were doing this in some other methods in the class. So, I did it to be consistent. But my own opinion is we don't really need it like this :)

File: readthedocs/core/utils/__init__.py (hunk 4 above)
Comment (saadmk11) [reply readthedocs/core/utils/__init__.py]: > Should this check for the `version` being external before sending?

@ericholscher I'm checking that inside the task
https://github.com/readthedocs/readthedocs.org/blob/0011043e85c3499c66079893806b06d1065aea84/readthedocs/projects/tasks.py#L1807-L1808

We are using this in few places so it did not look good to check everywhere so I put it here.

Comment (saadmk11) [reply readthedocs/core/utils/__init__.py]: > The status can be for any commit, I think it would be cool to report the status in normal versions

We don't store the `commit sha` of a branch.

File: readthedocs/projects/tasks.py (hunk 5 above)
Comment (saadmk11) [reply readthedocs/projects/tasks.py]: @ericholscher I'm checking that inside the task
https://github.com/readthedocs/readthedocs.org/blob/0011043e85c3499c66079893806b06d1065aea84/readthedocs/projects/tasks.py#L1807-L1808

We are using this in few places so it did not look good to check everywhere so I put it here.

File: readthedocs/core/utils/__init__.py [hunk 9]
```diff
@@ -125,6 +128,10 @@ def prepare_build(
...
+        # Send pending Build Status to Git Status API
+        send_build_status.delay(build.id, BUILD_STATUS_PENDING)
```
[the remark this answers is not in the document]
Comment (saadmk11) [reply readthedocs/core/utils/__init__.py]: @stsewd I couldn't not find a better place for this . we should send the status report as soon as possible. please suggest me a better place when you get time :)

File: readthedocs/oauth/services/github.py [hunk 10]
```diff
@@ -311,6 +312,83 @@ def update_webhook(self, project, integration):
             )
             return (False, resp)
 
+    def send_build_status(self, project, identifier, state, build_url):
```
[the remark this answers is not in the document]
Comment (saadmk11) [reply readthedocs/oauth/services/github.py]: Sorry! forgot to change it :)

File: readthedocs/oauth/services/github.py [hunk 11]
```diff
@@ -311,6 +312,83 @@ def update_webhook(self, project, integration):
...
+            "target_url": build_url,
+            "description": description,
+            "context": "continuous-documentation/read-the-docs"
+        }
+
+        resp = None
+        try:
+            resp = session.post(
+                (
+                    'https://api.github.com/repos/{owner}/{repo}/statuses/{sha}'
```
[the remark this answers is not in the document]
Comment (saadmk11) [reply readthedocs/oauth/services/github.py]: yes. but I was trying to be consistent with the other code of this class :)

File: readthedocs/projects/tasks.py (hunk 6 above)
Comment (saadmk11) [reply readthedocs/projects/tasks.py]: Thanks! I did not know this :100:

File: readthedocs/oauth/services/github.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/oauth/services/github.py]: We are actually using f-strings in new code

File: readthedocs/builds/models.py (hunk 2 above)
Comment (saadmk11) [reply readthedocs/builds/models.py]: Do you have any suggestions on which constant should I use?

Comment (ericholscher) [reply readthedocs/builds/models.py]: Looks like we already just assume `https` in the code: https://github.com/readthedocs/readthedocs.org/blob/f886e68260d1dd37135fb20278b40502f9b7d7b0/readthedocs/oauth/services/github.py#L172-L182

File: readthedocs/core/utils/__init__.py (hunk 4 above)
Comment (ericholscher) [reply readthedocs/core/utils/__init__.py]: > We are using this in few places so it did not look good to check everywhere so I put it here.

Yea, but it creates an entire task in order to be thrown away, which feels quite wasteful. I guess it's OK for now, but I'd really prefer to not have all those extra tasks triggered on normal builds.

File: readthedocs/builds/models.py (hunk 2 above)
Comment (saadmk11) [reply readthedocs/builds/models.py]: @ericholscher we wont be able to see the page when using `localhost` for testing as `localhost` does not support `https`

File: readthedocs/oauth/services/github.py [hunk 12]
```diff
@@ -311,6 +312,74 @@ def update_webhook(self, project, integration):
...
+                return True
+
+            if resp.status_code in [401, 403, 404]:
+                log.info(
+                    'GitHub project does not exist or user does not have '
+                    'permissions: project=%s',
+                    project,
+                )
+                return False
+
```
Comment (stsewd) [inline readthedocs/oauth/services/github.py]: This should return False in the last line of the try block, otherwise it will return None

File: readthedocs/core/utils/__init__.py (hunk 4 above)
Comment (saadmk11) [reply readthedocs/core/utils/__init__.py]: @ericholscher can we create a function that will check the version type and then call the task? or should we check for the type everywhere we use it?

File: readthedocs/oauth/services/github.py [hunk 13]
```diff
@@ -311,6 +312,74 @@ def update_webhook(self, project, integration):
...
+            log.exception(
+                'GitHub commit status creation failed for project: %s',
+                project,
+            )
+            # Response data should always be JSON, still try to log if not
+            # though
+            try:
+                debug_data = resp.json()
+            except ValueError:
+                debug_data = resp.content
```
Comment (stsewd) [inline readthedocs/oauth/services/github.py]: if session.post fails, resp will be None, and give an error here.

File: readthedocs/oauth/services/github.py [hunk 14]
```diff
@@ -311,6 +312,74 @@ def update_webhook(self, project, integration):
...
+        data = {
+            'state': github_build_state,
+            'target_url': build.get_full_url(),
+            'description': description,
+            'context': 'continuous-documentation/read-the-docs'
+        }
+
+        resp = None
+        try:
+
```
Comment (stsewd) [inline readthedocs/oauth/services/github.py]: extra space :man_astronaut:

File: readthedocs/oauth/services/github.py [hunk 15]
```diff
@@ -311,6 +312,79 @@ def update_webhook(self, project, integration):
...
+            log.exception(
+                'GitHub commit status creation failed for project: %s',
+                project,
+            )
+            # Response data should always be JSON, still try to log if not
+            # though
+            try:
+                debug_data = resp.json()
+            except ValueError:
+                if resp is not None:
```
Comment (stsewd) [inline readthedocs/oauth/services/github.py]: The if should go before the try block, otherwise `resp.json` will give `AttributeError` too.

Comment (saadmk11) [reply readthedocs/oauth/services/github.py]: yeah! I'm doing that now. I wonder how other methods are working with this :)

File: readthedocs/builds/constants.py (hunk 1 above)
Comment (ericholscher) [reply readthedocs/builds/constants.py]: Yes.

File: readthedocs/builds/models.py [hunk 16]
```diff
@@ -734,6 +734,14 @@ def __str__(self):
     def get_absolute_url(self):
         return reverse('builds_detail', args=[self.project.slug, self.pk])
 
+    def get_full_url(self):
+        """Get full url including domain"""
+        full_url = 'https://{domain}{absolute_url}'.format(
```
Comment (ericholscher) [inline readthedocs/builds/models.py]: We could check for DEBUG here and set it to `http` if we wanted something that works in dev, but I'd prefer a better solution. We already do this somewhere else: https://github.com/readthedocs/readthedocs.org/blob/master/readthedocs/api/v3/serializers.py#L31

File: readthedocs/doc_builder/environments.py [hunk 17]
```diff
@@ -658,7 +658,7 @@ def failed(self):
     def done(self):
         """Is build in finished state."""
         return (
-            self.build is not None and
+            self.build and
```
Comment (ericholscher) [inline readthedocs/doc_builder/environments.py]: Why was this changed?

File: readthedocs/oauth/services/base.py [hunk 18]
```diff
@@ -216,6 +216,9 @@ def setup_webhook(self, project):
     def update_webhook(self, project, integration):
         raise NotImplementedError
 
+    def send_build_status(self, build, state):
```
Comment (ericholscher) [inline readthedocs/oauth/services/base.py]: This should have a docstring. The one above it probably should too :)

File: readthedocs/projects/tasks.py [hunk 19]
```diff
@@ -573,6 +578,15 @@ def run_build(self, docker, record):
 
         if self.build_env.failed:
             self.send_notifications(self.version.pk, self.build['id'])
+            # send build failure status to git Status API
+            self.send_build_status(
+                self.build['id'], BUILD_STATUS_FAILURE
+            )
+        elif self.build_env.successful:
```
Comment (ericholscher) [inline readthedocs/projects/tasks.py]: What is the other state here? Should this just be an `else`? If not, we should do an `else` with a `log.warning`.

File: readthedocs/projects/tasks.py [hunk 20]
```diff
@@ -1008,6 +1022,10 @@ def is_type_sphinx(self):
         """Is documentation type Sphinx."""
         return 'sphinx' in self.config.doctype
 
+    def send_build_status(self, build_pk, state):
+        """Send github build status for pull/merge requests."""
+        send_external_build_status(build_pk, state)
```
Comment (ericholscher) [inline readthedocs/projects/tasks.py]: Do we need this function? Seems we can just call it directly.

File: readthedocs/projects/tasks.py [hunk 21]
```diff
@@ -1773,3 +1794,45 @@ def retry_domain_verification(domain_pk):
...
+    except RemoteRepository.DoesNotExist:
+        log.info('Remote repository does not exist for %s', build.project)
+
+    except Exception:
+        log.exception('Send build status task failed for %s', build.project)
+
+    # TODO: Send build status for other providers.
+
+
+def send_external_build_status(build_pk, state):
```
Comment (ericholscher) [inline readthedocs/projects/tasks.py]: Shouldn't this function be in `utils` instead of `tasks`?

File: readthedocs/doc_builder/environments.py (hunk 17 above)
Comment (saadmk11) [reply readthedocs/doc_builder/environments.py]: Here we initialize build as an empty dict which is not equal to `None` and it raises some errors while calling `successful` as `self.build is not None`  returns `True` and tries to get the key `state`
https://github.com/readthedocs/readthedocs.org/blob/801f63be924266af69a6bd4a9234a8010ebcb4b1/readthedocs/projects/tasks.py#L322

File: readthedocs/projects/tasks.py (hunk 19 above)
Comment (saadmk11) [reply readthedocs/projects/tasks.py]: @ericholscher there are no other states (`BuildEnvironment`) , it will succeed or fail (in the Build Environment) at this point of the build. we could add an warning if we want.

File: readthedocs/projects/tasks.py (hunk 20 above)
Comment (saadmk11) [reply readthedocs/projects/tasks.py]: I was just following Notification pattern to make the code consistent. but I guess we don't need it :)

File: readthedocs/projects/tasks.py (hunk 21 above)
Comment (saadmk11) [reply readthedocs/projects/tasks.py]: This will just create a circular import error, as the `send_build_status` task is already in the `task.py` file

File: readthedocs/rtd_tests/tests/test_celery.py [hunk 22]
```diff
@@ -329,3 +332,32 @@ def test_fileify_logging_when_wrong_version_pk(self, mock_logger):
...
+    @patch('readthedocs.projects.tasks.GitHubService.send_build_status')
+    def test_send_build_status_task_without_remote_repo(self, send_build_status, mock_logger):
+        external_version = get(Version, project=self.project, type=EXTERNAL)
+        external_build = get(
+            Build, project=self.project, version=external_version
+        )
+        tasks.send_build_status(external_build, BUILD_STATUS_SUCCESS)
+
+        send_build_status.assert_not_called()
+        mock_logger.info.assert_called_with(
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_celery.py]: I don't think mocking the logger is needed for this test

File: readthedocs/rtd_tests/tests/test_oauth.py [hunk 23]
```diff
@@ -142,6 +150,51 @@ def test_multiple_users_same_repo(self):
...
+    def test_send_build_status_value_error(self, session, mock_logger):
+        session().post.side_effect = ValueError
+        success = self.service.send_build_status(
+            self.external_build, BUILD_STATUS_SUCCESS
+        )
+
+        self.assertFalse(success)
+        mock_logger.exception.assert_called_with(
+                'GitHub commit status creation failed for project: %s',
+                self.project,
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_oauth.py]: Looks like there is an extra level of indentation here

**Comment by ericholscher** [comment]
:tada:
